src/charm.py

```diff
@@ -249,52 +263,96 @@ def _configure(self, _):
         line arguments). If the Pebble layer has changed then Prometheus
         is restarted.
         """
-        container = self.unit.get_container(self._name)
+        early_return_statuses = {
+            "cfg_load_fail": BlockedStatus(
+                "Prometheus failed to reload the configuration (WAL replay or ingress in progress?); see debug logs"
+            ),
+            "restart_fail": BlockedStatus(
+                "Prometheus failed to restart (config valid?); see debug logs"
+            ),
+            "push_fail": BlockedStatus(
+                "Failed to push updated config/alert files; see debug logs"
+            ),
+            "layer_fail": BlockedStatus("Failed to update prometheus service; see debug logs"),
+            "config_invalid": BlockedStatus("Invalid prometheus configuration; see debug logs"),
+            "validation_fail": BlockedStatus(
+                "Failed to validate prometheus config; see debug logs"
+            ),
+        }
```

Comment:
I like this. It seems slightly out of place in a Goldilocks way, though.

If we're going to have constant strings this way, they could be class-level rather than instance level, or just plain ALL-CAPS constants the way the old status was (just to make it clear what they are).

Alternatively, this seems like a great place for a tiny meta-config class/method, where:

```python
if some_bad_thing:
  self.set_unit_status("push_fail") # or even an Enum at the top to make it self.set_unit_status(PUSH_FAIL)
```

Would be a nice abstraction which keeps the list of conditions->statuses in a single place, almost like a `excepthook`

Reply:
Yep. But this would be a larger refactoring (a good one!) that we should probably discuss separately.